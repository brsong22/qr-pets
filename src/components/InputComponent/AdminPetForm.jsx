import React from 'react';
import Button from '@material-ui/core/Button';
import axios from 'axios';
import Upload from '../Upload';
import AdminTextInput from './AdminTextInput';

class AdminPetForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      hasChanged: false,
      fileList: [],
      tagList: [],
    };
    this.handleFileSelect = this.handleFileSelect.bind(this);
    this.handleTagsUpdate = this.handleTagsUpdate.bind(this);
    this.handleTextChange = this.handleTextChange.bind(this);
    this.saveForm = this.saveForm.bind(this);
  }

  handleFileSelect(files) {
    this.setState({
      fileList: files,
      hasChanged: true,
    });
  }

  handleTagsUpdate(tagsString) {
    const tags = tagsString.split(',').map(tag => tag.trim()).filter(tag => tag !== '');
    this.setState({
      tagList: tags,
      hasChanged: true,
    });
  }

  handleTextChange(changed) {
    this.setState({
      hasChanged: changed,
    });
  }

  async saveForm() {
    const { tagList, fileList } = this.state;
    const file = fileList[0];
    const { name, type } = file;
    const tags = tagList.map(tag => `${tag}&`).join(''); // values as tag keys with no value
    const signedUrl = await axios.post('/upload', { name, type });

    await axios.put(signedUrl.data, file, { headers: { 'x-amz-tagging': tags } });
  }

  render() {
    const { hasChanged } = this.state;
    return (
      <div>
        <Upload handleFileSelect={this.handleFileSelect} />
        <AdminTextInput handleUpdate={this.handleTagsUpdate} handleChange={this.handleTextChange} />
        <Button
          color="primary"
          disabled={!hasChanged}
          onClick={this.saveForm}
          variant="contained"
        >
          {'Save Form'}
        </Button>
      </div>
    );
  }
}

export default AdminPetForm;
