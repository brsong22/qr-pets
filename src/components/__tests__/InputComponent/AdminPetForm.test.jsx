import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import axios from 'axios';
import Button from '@material-ui/core/Button';
import AdminPetForm from '../../InputComponent/AdminPetForm';

jest.mock('axios', () => ({
  post: jest.fn().mockReturnValue({ data: 'someUrl' }),
  put: jest.fn().mockReturnValue({ data: ['something'] }),
}));

describe('AdminPetForm', () => {
  const file = { name: 'moomoo', type: 'html/jpeg' };
  const tagArray = ['guinea pig', 'moomoo', 'fluffy'];
  const firstPicture = { name: 'frodo', type: 'html/jpeg' };
  let wrapper;
  let wrapperInst;

  beforeEach(() => {
    wrapper = shallow(<AdminPetForm />);
    wrapperInst = wrapper.instance();
  });

  it('renders as expected', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('setState {image list, hasChanged} on handlefileSelect', () => {
    const img = [file];
    const fileSelectSpy = jest.spyOn(wrapperInst, 'handleFileSelect');
    wrapperInst.handleFileSelect(img);
    expect(fileSelectSpy).toHaveBeenCalledTimes(1);
    expect(wrapper.state().fileList).toEqual(img);
    expect(wrapper.state().hasChanged).toEqual(true);
  });

  it('setState {tagList, hasChanged} on handleTagsUpdate', () => {
    const tagsString = 'guinea pig, moomoo, fluffy,';
    const tagsState = ['guinea pig', 'moomoo', 'fluffy'];
    const tagsUpdateSpy = jest.spyOn(wrapperInst, 'handleTagsUpdate');
    wrapperInst.handleTagsUpdate(tagsString);
    expect(tagsUpdateSpy).toHaveBeenCalledTimes(1);
    expect(wrapper.state().tagList).toEqual(tagsState);
    expect(wrapper.state().hasChanged).toEqual(true);
  });

  it('setState {hasChanged} on handleTextChange', () => {
    const textUpdateSpy = jest.spyOn(wrapperInst, 'handleTextChange');
    wrapperInst.handleTextChange(true);
    expect(textUpdateSpy).toHaveBeenCalledTimes(1);
    expect(wrapper.state().hasChanged).toEqual(true);
  });

  it('stores file from input to state', () => {
    const InputComponent = wrapper.find('#file-chooser');
    const files = [firstPicture];

    InputComponent.simulate('change', files);

    expect(wrapper.state().fileList).toEqual([firstPicture]);
  });

  it('button is disabled if no file is selected', () => {
    const ButtonComponent = wrapper.find(Button);

    expect(ButtonComponent.props().disabled).toEqual(true);
  });

  it('button is enabled if file is selected', () => {
    wrapper.setState({ fileList: [firstPicture] });
    const ButtonComponent = wrapper.find(Button);

    expect(ButtonComponent.props().disabled).toEqual(false);
  });

  it('sucessfully uploads to s3 with saveForm function', async () => {
    wrapper.setState({ fileList: [file], tagList: tagArray });
    const { name, type } = file;
    // const tags = tagArray.map(tag => `${tag}&`).join('');
    const tags = 'guinea pig&moomoo&fluffy';
    const headers = { headers: { 'x-amz-tagging': tags } };
    await wrapperInst.saveForm();

    expect(axios.post).toHaveBeenCalledWith('/upload', { name, type });
    expect(axios.put).toHaveBeenCalledWith('someUrl', file, headers);
  });
});
