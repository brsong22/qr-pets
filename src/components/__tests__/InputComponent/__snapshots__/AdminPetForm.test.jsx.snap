// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPetForm renders as expected 1`] = `
<div>
  <Upload
    handleFileSelect={[Function]}
  />
  <AdminTextInput
    handleChange={[Function]}
    handleUpdate={[Function]}
  />
  <WithStyles(Button)
    color="primary"
    disabled={true}
    onClick={[Function]}
    variant="contained"
  >
    Save Form
  </WithStyles(Button)>
</div>
`;
